### PR TITLE
Adding more field errors.

### PR DESCRIPTION
**Pull request by n3wscott**

### CODE REVIEW

File: apis/field_error.go [hunk 1]
```diff
@@ -108,3 +108,12 @@ func ErrInvalidValue(value string, fieldPath string) *FieldError {
 		Paths:   []string{fieldPath},
 	}
 }
+
+// ErrMutuallyExclusiveFields is a variadic helper method for constructing a FieldError for a set of mutually exclusive fields.
```
Comment (mattmoor) [inline apis/field_error.go]: nit: wrap so github code review doesn't make my eyes bleed.

File: apis/field_error.go [hunk 2]
```diff
@@ -108,3 +108,12 @@ func ErrInvalidValue(value string, fieldPath string) *FieldError {
...
+// Must have at least one set, but only one.
+func ErrMutuallyExclusiveFields(fieldPaths ...string) *FieldError {
+	return &FieldError{
+		Message: "mutually exclusive fields, must set only one of",
```
Comment (mattmoor) [inline apis/field_error.go]: There are actually multiple failure modes for this, we need this here: https://github.com/knative/serving/blob/5812169d86432d4895425a00966809223abc379c/pkg/apis/serving/v1alpha1/service_validation.go#L35-L49

Do you want to expose both?

Comment (mattmoor) [reply apis/field_error.go]: Perhaps `ErrMissingOneOf`, or `ErrMultipleOneOf`?

Former takes all options, latter takes just those specified.

File: apis/field_error_test.go [hunk 3]
```diff
@@ -101,7 +101,18 @@ Body.`,
 		err:      ErrInvalidValue("foo", "bar"),
 		prefixes: [][]string{{"baz"}},
 		want:     `invalid value "foo": baz.bar`,
-	}}
+	}, {
+		name:     "mutually exclusive fields",
+		err:      ErrMutuallyExclusiveFields("foo", "bar"),
+		prefixes: [][]string{{CurrentField}},
```
Comment (mattmoor) [inline apis/field_error_test.go]: LOL, `CurrentField` is not meant to be used like this, just as a starting point.

**Comment by n3wscott** [comment]
Updated, PTAL

File: apis/field_error.go [hunk 4]
```diff
@@ -85,26 +85,57 @@ func (fe *FieldError) Error() string {
...
+func ErrMultipleOneOf(fieldPaths ...string) *FieldError {
+	return &FieldError{
+		Message: "expected exactly one, got neither",
+		Paths:   fieldPaths,
+	}
+}
+
+// ErrInvalidKeyName is a variadic helper method for constructing a
+// FieldError that specifies a key name that is invalid.
+func ErrInvalidKeyName(value, fieldPath string, errs []string) *FieldError {
```
Comment (mattmoor) [inline apis/field_error.go]: make `errs ...error` or `err ...string`?

File: apis/field_error.go [hunk 5]
```diff
@@ -85,26 +85,57 @@ func (fe *FieldError) Error() string {
...
+	return &FieldError{
+		Message: fmt.Sprintf("invalid key name %q", value),
+		Paths:   []string{fieldPath},
+		Details: strings.Join(errs, ", "),
+	}
+}
```
Comment (mattmoor) [inline apis/field_error.go]: nit: add newline

**Comment by mattmoor** [comment]
/approve

will LGTM after comments

**Pull request merged**